File: backend/core/pagination.py
"""通用分页模型"""
from typing import Annotated, Generic, Sequence, TypeVar

from fastapi import Depends
from pydantic import BaseModel, Field

T = TypeVar("T")


class PageResponse(BaseModel, Generic[T]):
    items: Sequence[T]
    total: int
    page: int
    page_size: int
    total_pages: int = 0

    def __init__(self, **data):
        super().__init__(**data)
        self.total_pages = (self.total + self.page_size - 1) // self.page_size


class PaginationParams(BaseModel):
    page: int = Field(1, ge=1)
    page_size: int = Field(10, ge=1, le=100)

    @property
    def offset(self) -> int:
        return (self.page - 1) * self.page_size

    @property
    def limit(self) -> int:
        return self.page_size


PaginationDep = Annotated[PaginationParams, Depends()]

File: internal/storage/pagination.go
package storage

// PageResult 泛型分页响应，匹配 Python PageResponse 的语义。
type PageResult[T any] struct {
	Items      []T   `json:"items"`
	Total      int64 `json:"total"`
	Page       int   `json:"page"`
	Size       int   `json:"size"`
	TotalPages int   `json:"total_pages"`
}

// NewPageResult 根据 total/size 自动计算 TotalPages。
func NewPageResult[T any](items []T, total int64, page, size int) *PageResult[T] {
	tp := 0
	if size > 0 {
		tp = int(total) / size
		if int(total)%size != 0 {
			tp++
		}
	}
	return &PageResult[T]{
		Items:      items,
		Total:      total,
		Page:       page,
		Size:       size,
		TotalPages: tp,
	}
}
